File: TalentSystemDefaults/TalentSystemDefaults/DataAccess/DataObjects/TableObjects/tbl_ecommerce_module_defaults_bu.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Linq;
using System.Text;
using TalentSystemDefaults.DataEntities;
namespace TalentSystemDefaults
{
	namespace DataAccess.DataObjects
	{
		public class tbl_ecommerce_module_defaults_bu : DBObjectBase
		{
			#region Shared Properties
			static private bool _DefaultKey1Active = false;
			public static bool DefaultKey1Active
			{
				get
				{
					return _DefaultKey1Active;
				}
				set
				{
					_DefaultKey1Active = value;
				}
			}
			static private bool _DefaultKey2Active = false;
			public static bool DefaultKey2Active
			{
				get
				{
					return _DefaultKey2Active;
				}
				set
				{
					_DefaultKey2Active = value;
				}
			}
			static private bool _DefaultKey3Active = false;
			public static bool DefaultKey3Active
			{
				get
				{
					return _DefaultKey3Active;
				}
				set
				{
					_DefaultKey3Active = value;
				}
			}
			static private bool _DefaultKey4Active = false;
			public static bool DefaultKey4Active
			{
				get
				{
					return _DefaultKey4Active;
				}
				set
				{
					_DefaultKey4Active = value;
				}
			}
			static private bool _DefaultNameActive = true;
			public static bool DefaultNameActive
			{
				get
				{
					return _DefaultNameActive;
				}
				set
				{
					_DefaultNameActive = value;
				}
			}
			static private bool _VariableKey1Active = false;
			public static bool VariableKey1Active
			{
				get
				{
					return _VariableKey1Active;
				}
				set
				{
					_VariableKey1Active = value;
				}
			}
			static private bool _VariableKey2Active = false;
			public static bool VariableKey2Active
			{
				get
				{
					return _VariableKey2Active;
				}
				set
				{
					_VariableKey2Active = value;
				}
			}
			static private bool _VariableKey3Active = false;
			public static bool VariableKey3Active
			{
				get
				{
					return _VariableKey3Active;
				}
				set
				{
					_VariableKey3Active = value;
				}
			}
			static private bool _VariableKey4Active = false;
			public static bool VariableKey4Active
			{
				get
				{
					return _VariableKey4Active;
				}
				set
				{
					_VariableKey4Active = value;
				}
			}
			#endregion
			#region Class Level Fields
			const string TBL_ECOMMERCE_MODULE_DEFAULTS_BU = "tbl_ecommerce_module_defaults_bu";
			#endregion
			#region Constructors
			/// <summary>
			/// Constructor
			/// </summary>
			/// <param name="settings"></param>
			/// <remarks></remarks>
			public tbl_ecommerce_module_defaults_bu(ref DESettings settings)
				: base(settings)
			{
				businessUnit = settings.BusinessUnit;
			}
			#endregion
			#region Public Methods
			public override void RetrieveCurrentValues(System.Collections.Generic.List<ConfigurationItem> configs)
			{
				var currentValues = RetrieveCurrentValues(businessUnit, configs);
				AddMissingValues(configs, currentValues);
				string currentValue = "";
				foreach (ConfigurationItem config in configs)
				{
					if (currentValues.ContainsKey(config.DefaultName))
					{
						currentValue = currentValues[config.DefaultName];
						config.CurrentValue = currentValue;
						config.UpdatedValue = currentValue;
					}
				}
			}
			public override int UpdateCurrentValue(System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction = null)
			{
				return UpdateData(businessUnit, configs, givenTransaction);
			}
			public override List<ConfigurationEntity> RetrieveAllValues(string businessUnit, string[] defaultKeys, string[] variableKeys, string displayName = "")
			{
				DataTable outputDataTable = new DataTable();
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				string commandText = "SELECT distinct default_name,value FROM dbo.tbl_ecommerce_module_defaults_bu WHERE business_unit = @BUSINESS_UNIT" + (displayName != string.Empty ? " AND DEFAULT_NAME like \'%\' + @DEFAULT_NAME + \'%\'" : "");
				ErrorObj err = new ErrorObj();
				try
				{
					//Construct The Call
					talentSqlAccessDetail.Settings = settings;
					talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet;
					talentSqlAccessDetail.CommandElements.CommandParameter.Clear();
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
					if (displayName != string.Empty)
					{
						talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DEFAULT_NAME", displayName));
					}
					talentSqlAccessDetail.CommandElements.CommandText = commandText;
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
					if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
					{
						outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables[0];
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
				return GetConfigurationData(outputDataTable, defaultKeys, variableKeys);
			}
			/// <summary>
			/// Returns the group ID
			/// </summary>
			/// <returns></returns>
			/// <remarks></remarks>
			public int GetGroupID()
			{
				DataTable outputDataTable = new DataTable();
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				StringBuilder commandText = new StringBuilder();
				ErrorObj err = new ErrorObj();
				try
				{
					commandText.AppendLine("select value from dbo.tbl_ecommerce_module_defaults_bu a where a.DEFAULT_NAME = \'TALENT_DEFAULTS_GROUP_ID\'");
					commandText.AppendLine("update	dbo.tbl_ecommerce_module_defaults_bu");
					commandText.AppendLine("set		VALUE = cast(VALUE as int) + 1 ");
					commandText.AppendLine("where	DEFAULT_NAME = \'TALENT_DEFAULTS_GROUP_ID\'");
					//Construct The Call
					talentSqlAccessDetail.Settings = settings;
					talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet;
					talentSqlAccessDetail.CommandElements.CommandText = commandText.ToString();
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
					if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
					{
						outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables[0];
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
				return int.Parse(System.Convert.ToString(outputDataTable.Rows[0]["value"]));
			}
			/// <summary>
			/// Returns web API url from database
			/// </summary>
			/// <returns></returns>
			/// <remarks></remarks>
			public string GetSystemDefaultsURL()
			{
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				string commandText = "SELECT VALUE FROM DBO.TBL_ECOMMERCE_MODULE_DEFAULTS_BU WHERE BUSINESS_UNIT = @BUSINESS_UNIT AND DEFAULT_NAME = \'SYSTEM_DEFAULTS_URL\'";
				ErrorObj err = new ErrorObj();
				string retVal = string.Empty;
				try
				{
					//Construct The Call
					talentSqlAccessDetail.Settings = settings;
					talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet;
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
					talentSqlAccessDetail.CommandElements.CommandText = commandText;
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
					if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
					{
						System.Data.DataTable with_3 = talentSqlAccessDetail.ResultDataSet.Tables[0];
						if (with_3.Rows.Count > 0)
						{
							retVal = with_3.Rows[0]["VALUE"].ToString();
						}
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
				return retVal;
			}
			#endregion
			#region Private Methods
			/// <summary>
			/// Returns current configurations
			/// </summary>
			/// <param name="businessUnit"></param>
			/// <param name="configs"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private Dictionary<string, string> RetrieveCurrentValues(string businessUnit, System.Collections.Generic.List<ConfigurationItem> configs)
			{
				DataTable outputDataTable = new DataTable();
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				string[] whereClause = new string[configs.Count - 1 + 1];
				string where = string.Empty;
				string commandText = string.Empty;
				ErrorObj err = new ErrorObj();
				try
				{
					for (int i = 0; i <= configs.Count - 1; i++)
					{
						whereClause[i] = "(default_name = @DEFAULT_NAME" + i.ToString() + ")";
					}
					where = string.Join(" OR ", whereClause);
					commandText = string.Format("SELECT distinct ID,DEFAULT_NAME,VALUE FROM dbo.tbl_ecommerce_module_defaults_bu WHERE business_unit = @BUSINESS_UNIT AND ({0})", where);
					//Construct The Call
					talentSqlAccessDetail.Settings = settings;
					talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet;
					talentSqlAccessDetail.CommandElements.CommandParameter.Clear();
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
					for (int i = 0; i <= configs.Count - 1; i++)
					{
						talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DEFAULT_NAME" + i.ToString(), configs[i].DefaultName));
					}
					talentSqlAccessDetail.CommandElements.CommandText = commandText;
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
					if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
					{
						outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables[0];
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
				return GetCurrentValues(outputDataTable);
			}
			/// <summary>
			/// Returns set of configurations read from datatable passed
			/// </summary>
			/// <param name="dt"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private Dictionary<string, string> GetCurrentValues(DataTable dt)
			{
				var results = new Dictionary<string, string>();
				foreach (DataRow row in dt.Rows)
				{
					var attrName = Utilities.CheckForDBNull_String(row["DEFAULT_NAME"]).Trim();
					var attrValue = Utilities.CheckForDBNull_String(row["VALUE"]).Trim();
					if (!results.ContainsKey(attrName))
					{
						results.Add(attrName, attrValue);
					}
				}
				return results;
			}
			/// <summary>
			/// Returns configurations as list of ConfigurationEntity read from dataTable passed
			/// </summary>
			/// <param name="dt"></param>
			/// <param name="defaultKeys"></param>
			/// <param name="variableKeys"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private List<ConfigurationEntity> GetConfigurationData(DataTable dt, string[] defaultKeys, string[] variableKeys)
			{
				var results = new System.Collections.Generic.List<ConfigurationEntity>();
				foreach (DataRow row in dt.Rows)
				{
					var dispName = Utilities.CheckForDBNull_String(row["DEFAULT_NAME"]);
					var attrName = Utilities.CheckForDBNull_String(row["DEFAULT_NAME"]);
					var attrValue = Utilities.CheckForDBNull_String(row["VALUE"]);
					//Dim attrDescription = Utilities.CheckForDBNull_String(row("DESCRIPTION"))
					results.Add(new ConfigurationEntity(TBL_ECOMMERCE_MODULE_DEFAULTS_BU, string.Empty, defaultKeys, variableKeys, dispName, attrName, attrValue, string.Empty, string.Empty, ""));
				}
				return results;
			}
			/// <summary>
			/// Updates configurations
			/// </summary>
			/// <param name="businessUnit"></param>
			/// <param name="configs"></param>
			/// <param name="givenTransaction"></param>
			/// <returns></returns>
			/// <remarks></remarks>
			private int UpdateData(string businessUnit, System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction)
			{
				StringBuilder cases = new StringBuilder();
				for (int i = 0; i <= configs.Count - 1; i++)
				{
					cases.Append(" WHEN  @DEFAULT_NAME" + i.ToString() + " THEN  @DEFAULT_VALUE" + i.ToString());
				}
				string[] whereClause = new string[configs.Count - 1 + 1];
				for (int i = 0; i <= configs.Count - 1; i++)
				{
					whereClause[i] = "(DEFAULT_NAME = @DEFAULT_NAME" + i.ToString() + ")";
				}
				string where = string.Join(" OR ", whereClause);
				var commandText = string.Format("UPDATE dbo.tbl_ecommerce_module_defaults_bu SET VALUE = CASE default_name {0} END  WHERE business_unit = @BUSINESS_UNIT AND ({1})", cases.ToString(), where);
				int affectedRows = 0;
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				ErrorObj err = new ErrorObj();
				//Construct The Call
				talentSqlAccessDetail.Settings = settings;
				talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery;
				talentSqlAccessDetail.CommandElements.CommandText = commandText;
				talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
				for (int i = 0; i <= configs.Count - 1; i++)
				{
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DEFAULT_NAME" + i.ToString(), configs[i].DefaultName));
					talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DEFAULT_VALUE" + i.ToString(), configs[i].UpdatedValue));
				}
				//Execute
				if (givenTransaction == null)
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
				}
				else
				{
					err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005, givenTransaction);
				}
				if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
				{
					affectedRows = System.Convert.ToInt32(talentSqlAccessDetail.ResultDataSet.Tables[0].Rows[0][0]);
				}
				talentSqlAccessDetail = null;
				//Return results
				return affectedRows;
			}

			public override void AddMissingValues(System.Collections.Generic.List<ConfigurationItem> configs, Dictionary<string, string> currentValues)
			{
				List<ConfigurationItem> missingConfigs;
				DataTable outputDataTable = new DataTable();
				TalentDataAccess talentSqlAccessDetail = new TalentDataAccess();
				talentSqlAccessDetail.Settings = settings;
				talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery;
				ErrorObj err = new ErrorObj();
				List<string> items = new List<string>();
				string commandText = String.Empty;
				string values = string.Empty;
				try
				{
					missingConfigs = configs.Where(c => !currentValues.ContainsKey(c.DefaultName)).ToList();
					if (missingConfigs.Count > 0)
					{
						talentSqlAccessDetail.CommandElements.CommandParameter.Clear();
						talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BUSINESS_UNIT", businessUnit));
						talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@PARTNER_CODE", ALL_STRING));
						talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@APPLICATION", string.Empty));
						talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@MODULE", string.Empty));
						for (int i = 0; i < missingConfigs.Count; i++)
						{
							int j = i + 1;
							ConfigurationItem config = missingConfigs[i];
							String text = string.Format("SELECT @BUSINESS_UNIT, @PARTNER_CODE, @APPLICATION, @MODULE, @DEFAULT_NAME_{0}, @VALUE_{0}", j);
							talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DEFAULT_NAME_" + j, config.DefaultName));
							talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@VALUE_" + j, config.CurrentValue));
							items.Add(text);
						}

						values = String.Join(" UNION ", items.ToArray());
						commandText = string.Format("INSERT into tbl_ecommerce_module_defaults_bu {0}", values);
						talentSqlAccessDetail.CommandElements.CommandText = commandText;

						err = talentSqlAccessDetail.SQLAccess(DestinationDatabase.SQL2005);
						if ((!(err.HasError)) && (!(talentSqlAccessDetail.ResultDataSet == null)))
						{
							outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables[0];
						}
					}
				}
				catch (Exception)
				{
					throw;
				}
				finally
				{
					talentSqlAccessDetail = null;
				}
			}
			#endregion
		}
	}
}

File: TalentCommon/DataObjects/ApplicationVariables/tbl_ecommerce_module_defaults_bu.vb
Imports System.Data.SqlClient

' This Enumeration is outside the namespace DataObjects to make sure that it is accessed 
' whereever Talent.Common namespace is refered
#Region "Enumerations"
''' <summary>
''' Each memeber in this enum is value for the column DEFAULT_NAME in the table tbl_ecommerce_module_defaults_bu
''' </summary>
Public Enum TblEcommerceModuleDefaultsBuEnum
    ''' <summary>
    ''' Abolute path for cache
    ''' ex: C:\foldername\foldername\.....
    ''' </summary>
    CACHE_DEPENDENCY_PATH
    ''' <summary>
    ''' Valid email address
    ''' </summary>
    CONTACT_US_FROM_EMAIL
    ''' <summary>
    ''' Valid email address for receipient
    ''' </summary>
    CONTACT_US_TO_EMAIL
    ''' <summary>
    ''' Valid email address for cc
    ''' </summary>
    CONTACT_US_TO_EMAIL_CC
    ''' <summary>
    ''' Valid email address
    ''' </summary>
    CUSTOMER_SERVICES_EMAIL
    ''' <summary>
    ''' Valid email address
    ''' </summary>
    DEFAULT_TICKETING_EMAIL_ADDRESS
    ''' <summary>
    ''' New Business Unit Stadium name
    ''' </summary>
    FLASH_STADIUM_NAME
    ''' <summary>
    ''' Abolute path for html files
    ''' ex: C:\foldername\foldername\.....
    ''' </summary>
    HTML_PATH_ABSOLUTE
    ''' <summary>
    ''' Abolute path for image files
    ''' ex: C:\foldername\foldername\.....
    ''' </summary>
    IMAGE_PATH_ABSOLUTE
    ''' <summary>
    ''' key for encrypting the NOISE service
    ''' </summary>
    NOISE_ENCRYPTION_KEY
    ''' <summary>
    ''' URL for NOISE service
    ''' </summary>
    NOISE_URL
    ''' <summary>
    ''' Valid email address
    ''' </summary>
    REGISTRATION_CONFIRMATION_FROM_EMAIL
    ''' <summary>
    ''' valid group name
    ''' </summary>
    STORED_PROCEDURE_GROUP
    ''' <summary>
    ''' valid theme name
    ''' </summary>
    THEME
    ''' <summary>
    ''' Ticketing stadium name
    ''' </summary>
    TICKETING_STADIUM
    ''' <summary>
    ''' Authority user profile - Backend Internet user
    ''' </summary>
    AUTHORITY_USER_PROFILE
    ''' <summary>
    ''' Payment 3D Secure Attrubute 1
    ''' </summary>
    PAYMENT_3DSECURE_DETAILS_1
    ''' <summary>
    ''' Payment 3D Secure Attrubute 2
    ''' </summary>
    PAYMENT_3DSECURE_DETAILS_2
    ''' <summary>
    ''' Payment 3D Secure Attrubute 3
    ''' </summary>
    PAYMENT_3DSECURE_DETAILS_3
    ''' <summary>
    ''' Payment 3D Secure, to pass the enrollment check if the return code is U (boolean value)
    ''' </summary>
    PAYMENT_3DSECURE_PASS_IF_ENROLLED_CODE_U
    ''' <summary>
    ''' Payment 3D Secure, to pass the enrollment check if the response times out (boolean value)
    ''' </summary>
    PAYMENT_3DSECURE_PASS_IF_RESPONSE_TIMEOUT
End Enum
#End Region

Namespace DataObjects.TableObjects

    ''' <summary>
    ''' Provides the functionality to manage data from the table tbl_ecommerce_module_defaults_bu based on business functionality
    ''' </summary>
    <Serializable()> _
    Public Class tbl_ecommerce_module_defaults_bu
        Inherits DBObjectBase

#Region "Class Level Fields"

        ''' <summary>
        ''' Instance of DESettings
        ''' </summary>
        Private _settings As New DESettings

        ''' <summary>
        ''' Class Name which is used in cache key construction
        ''' </summary>
        Const CACHEKEY_CLASSNAME As String = "tbl_ecommerce_module_defaults_bu"

        ''' <summary>
        ''' To decide to continue the loop of dictionary object in UpdateMultiple method
        ''' </summary>
        Private _continueUpdateMultiple As Boolean = True

#End Region

#Region "Constructors"
        Sub New()
        End Sub
        ''' <summary>
        ''' Initializes a new instance of the <see cref="tbl_ecommerce_module_defaults_bu" /> class.
        ''' </summary>
        ''' <param name="settings">The DESettings instance</param>
        Sub New(ByVal settings As DESettings)
            _settings = settings
        End Sub
#End Region

#Region "Public Methods"

        ''' <summary>
        ''' Deletes the specified business unit.
        ''' </summary>
        ''' <param name="businessUnitToDelete">The business unit to delete.</param>
        ''' <param name="givenTransaction">The given transaction.</param>
        ''' <returns>No of affected rows</returns>
        Public Function Delete(ByVal businessUnitToDelete As String, Optional ByVal givenTransaction As SqlTransaction = Nothing) As Integer

            Dim affectedRows As Integer = 0
            Dim talentSqlAccessDetail As New TalentDataAccess
            Dim err As New ErrorObj
            'Construct The Call
            talentSqlAccessDetail.Settings = _settings
            talentSqlAccessDetail.Settings.Cacheing = False
            talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery
            Dim sqlStatement As String = "DELETE TBL_ECOMMERCE_MODULE_DEFAULTS_BU " & _
                "WHERE BUSINESS_UNIT=@BusinessUnitToDelete "
            talentSqlAccessDetail.CommandElements.CommandText = sqlStatement
            talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitToDelete", businessUnitToDelete))

            'Execute
            If (givenTransaction Is Nothing) Then
                err = talentSqlAccessDetail.SQLAccess()
            Else
                err = talentSqlAccessDetail.SQLAccess(givenTransaction)
            End If
            If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                affectedRows = talentSqlAccessDetail.ResultDataSet.Tables(0).Rows(0)(0)
            End If
            talentSqlAccessDetail = Nothing

            'Return results
            Return affectedRows

        End Function

        ''' <summary>
        ''' Copies the records from one business unit to another business unit inside the table tbl_ecommerce_module_defaults_bu
        ''' irrespective of whether destination business exists or not and returns no of affected rows
        ''' </summary>
        ''' <param name="fromBusinessUnit">From business unit.</param>
        ''' <param name="toBusinessUnit">To business unit.</param>
        ''' <returns>No Of Affected Rows</returns>
        Public Function CopyByBU(ByVal fromBusinessUnit As String, ByVal toBusinessUnit As String, Optional ByVal givenTransaction As SqlTransaction = Nothing) As Integer

            Dim affectedRows As Integer = 0
            Dim talentSqlAccessDetail As New TalentDataAccess
            Try
                'Construct The Call

                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = False
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery

                Dim sqlStatement As String = "INSERT INTO TBL_ECOMMERCE_MODULE_DEFAULTS_BU (" & _
                    "BUSINESS_UNIT, APPLICATION, DEFAULT_NAME, " & _
                    "MODULE, PARTNER, VALUE) " & _
                    "SELECT " & _
                    "@ToBusinessUnit As BUSINESS_UNIT, APPLICATION, DEFAULT_NAME, " & _
                    "MODULE, PARTNER, VALUE " & _
                    "FROM TBL_ECOMMERCE_MODULE_DEFAULTS_BU " & _
                    "WHERE BUSINESS_UNIT = @FromBusinessUnit"

                talentSqlAccessDetail.CommandElements.CommandText = sqlStatement
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@FromBusinessUnit", fromBusinessUnit))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@ToBusinessUnit", toBusinessUnit))

                'Execute
                Dim err As New ErrorObj
                If (givenTransaction Is Nothing) Then
                    err = talentSqlAccessDetail.SQLAccess()
                Else
                    err = talentSqlAccessDetail.SQLAccess(givenTransaction)
                End If
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    affectedRows = talentSqlAccessDetail.ResultDataSet.Tables(0).Rows(0)(0)
                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return affectedRows

        End Function

        ''' <summary>
        ''' Gets the business unit records based on the given business unit.
        ''' </summary>
        ''' <param name="businessUnit">The business unit.</param>
        ''' <param name="cacheing">if set to <c>true</c> [cacheing]. Default <c>true</c></param>
        ''' <param name="cacheTimeMinutes">The cache time minutes. Default <c>30</c></param>
        ''' <returns>DataTable</returns>
        Public Function GetByBU(ByVal businessUnit As String, Optional ByVal cacheing As Boolean = True, Optional ByVal cacheTimeMinutes As Integer = 30) As DataTable

            Dim outputDataTable As New DataTable
            Dim cacheKeyPrefix As String = GetCacheKeyPrefix(CACHEKEY_CLASSNAME, "GetByBU")
            Dim talentSqlAccessDetail As New TalentDataAccess

            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                talentSqlAccessDetail.Settings.CacheStringExtension = cacheKeyPrefix & ToUpper(businessUnit)
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet
                talentSqlAccessDetail.CommandElements.CommandText = "Select * From TBL_ECOMMERCE_MODULE_DEFAULTS_BU Where BUSINESS_UNIT=@BusinessUnitName"
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitName", businessUnit))

                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables(0)
                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return outputDataTable

        End Function

        Public Function GetDefaultValues(ByVal partner As String, ByVal businessUnit As String, Optional ByVal cacheing As Boolean = True, Optional ByVal cacheTimeMinutes As Integer = 30) As DataTable

            Dim outputDataTable As New DataTable
            Dim cacheKeyPrefix As String = GetCacheKeyPrefix(CACHEKEY_CLASSNAME, "GetByBU" & partner)
            Dim talentSqlAccessDetail As New TalentDataAccess

            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                talentSqlAccessDetail.Settings.CacheStringExtension = cacheKeyPrefix & ToUpper(businessUnit)
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet
                talentSqlAccessDetail.CommandElements.CommandText = "Select * From TBL_ECOMMERCE_MODULE_DEFAULTS_BU Where BUSINESS_UNIT=@BusinessUnitName AND PARTNER=@Partner"
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitName", businessUnit))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@Partner", partner))

                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables(0)
                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return outputDataTable

        End Function
        ''' <summary>
        ''' Gets a single VALUE from  the given BUSINESS_UNIT and DEFAULT_NAME.
        ''' </summary>
        ''' <param name="businessUnit">The business unit.</param>
        ''' <param name="cacheing">if set to <c>true</c> [cacheing]. Default <c>true</c></param>
        ''' <param name="cacheTimeMinutes">The cache time minutes. Default <c>30</c></param>
        ''' <returns>DataTable</returns>
        Public Function GetValueByDefaultNameAndBU(ByVal businessUnit As String, ByVal defaultName As String, Optional ByVal cacheing As Boolean = True, Optional ByVal cacheTimeMinutes As Integer = 30) As String

            Dim value As String = String.Empty
            Dim cacheKeyPrefix As String = GetCacheKeyPrefix(CACHEKEY_CLASSNAME, "GetValueByDefaultNameAndBU")
            Dim talentSqlAccessDetail As New TalentDataAccess

            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                talentSqlAccessDetail.Settings.CacheStringExtension = cacheKeyPrefix & ToUpper(businessUnit) & defaultName.ToUpper
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet
                talentSqlAccessDetail.CommandElements.CommandText = "SELECT TOP 1 [VALUE] FROM [TBL_ECOMMERCE_MODULE_DEFAULTS_BU] WHERE [BUSINESS_UNIT] = @BusinessUnitName AND [DEFAULT_NAME] = @DefaultName"
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitName", businessUnit))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DefaultName", defaultName))

                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    If talentSqlAccessDetail.ResultDataSet.Tables.Count > 0 Then
                        If talentSqlAccessDetail.ResultDataSet.Tables(0).Rows.Count > 0 Then
                            value = talentSqlAccessDetail.ResultDataSet.Tables(0).Rows(0).Item(0).ToString
                        End If
                    End If

                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return value

        End Function

        Public Function GetDefaultNameValue(ByVal businessUnit As String, ByVal partner As String, ByVal defaultName As String, Optional ByVal cacheing As Boolean = True, Optional ByVal cacheTimeMinutes As Integer = 30) As String

            Dim value As String = String.Empty
            Dim cacheKeyPrefix As String = GetCacheKeyPrefix(CACHEKEY_CLASSNAME, "GetDefaultNameValue")
            Dim talentSqlAccessDetail As New TalentDataAccess

            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                talentSqlAccessDetail.Settings.CacheStringExtension = cacheKeyPrefix & ToUpper(businessUnit) & ToUpper(partner) & defaultName.ToUpper
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteDataSet
                talentSqlAccessDetail.CommandElements.CommandText = "SELECT TOP 1 [VALUE] FROM [TBL_ECOMMERCE_MODULE_DEFAULTS_BU] WHERE [BUSINESS_UNIT] = @BusinessUnitName AND [PARTNER] = @Partner AND [DEFAULT_NAME] = @DefaultName"
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnitName", businessUnit))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@Partner", partner))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DefaultName", defaultName))

                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    If talentSqlAccessDetail.ResultDataSet.Tables.Count > 0 Then
                        If talentSqlAccessDetail.ResultDataSet.Tables(0).Rows.Count > 0 Then
                            value = talentSqlAccessDetail.ResultDataSet.Tables(0).Rows(0).Item(0).ToString
                        End If
                    End If

                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return value

        End Function

        ''' <summary>
        ''' Updates the value for the specified business unit and default name
        ''' </summary>
        ''' <param name="businessUnit">The business unit.</param>
        ''' <param name="defaultName">The default name.</param>
        ''' <param name="defaultValue">The default value.</param>
        ''' <param name="givenTransaction">The given transaction.</param>
        ''' <returns>No of affected rows</returns>
        Public Function UpdateByBUAndDefaultName(ByVal businessUnit As String, ByVal defaultName As String, ByVal defaultValue As String, Optional ByVal givenTransaction As SqlTransaction = Nothing) As Integer

            Dim affectedRows As Integer = 0
            Dim talentSqlAccessDetail As New TalentDataAccess

            Try
                'Construct The Call
                talentSqlAccessDetail.Settings = _settings
                talentSqlAccessDetail.Settings.Cacheing = False
                talentSqlAccessDetail.CommandElements.CommandExecutionType = CommandExecution.ExecuteNonQuery
                Dim sqlStatement As String = "UPDATE TBL_ECOMMERCE_MODULE_DEFAULTS_BU " & _
                    "SET [VALUE] = @DefaultValue " & _
                    "WHERE BUSINESS_UNIT=@BusinessUnit AND DEFAULT_NAME = @DefaultName"

                talentSqlAccessDetail.CommandElements.CommandText = sqlStatement
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@BusinessUnit", businessUnit))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DefaultName", defaultName))
                talentSqlAccessDetail.CommandElements.CommandParameter.Add(ConstructParameter("@DefaultValue", defaultValue))


                'Execute
                Dim err As New ErrorObj
                If (givenTransaction Is Nothing) Then
                    err = talentSqlAccessDetail.SQLAccess()
                Else
                    err = talentSqlAccessDetail.SQLAccess(givenTransaction)
                End If
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    affectedRows = talentSqlAccessDetail.ResultDataSet.Tables(0).Rows(0)(0)
                Else
                    ' this is to decide whether to continue the for loop in updatemultiple method
                    'if any error exit the for loop
                    _continueUpdateMultiple = False
                End If
            Catch ex As Exception
                Throw
            Finally
                talentSqlAccessDetail = Nothing
            End Try

            'Return the results 
            Return affectedRows

        End Function

        ''' <summary>
        ''' Updates the default name of the multiple by BU and.
        ''' </summary>
        ''' <param name="businessUnit">The business unit.</param>
        ''' <param name="defaultNamesAndValues">The default names and values generic dictionary collection</param>
        ''' <param name="givenTransaction">The given transaction.</param>
        ''' <returns>total number of affected rows</returns>
        Public Function UpdateMultipleByBUAndDefaultName(ByVal businessUnit As String, ByVal defaultNamesAndValues As Dictionary(Of TblEcommerceModuleDefaultsBuEnum, String), Optional ByVal givenTransaction As SqlTransaction = Nothing) As Integer
            Dim totalAffectedRows As Integer = 0

            'Loops the dictionary if exists with key value pairs
            If Not (defaultNamesAndValues Is Nothing) Then
                If (defaultNamesAndValues.Count > 0) Then
                    _continueUpdateMultiple = True
                    Dim keyValue As KeyValuePair(Of TblEcommerceModuleDefaultsBuEnum, String)
                    Dim affectedRows As Integer = 0
                    For Each keyValue In defaultNamesAndValues
                        'to make sure there is no error in the UpdateByBUAndDefaultName method
                        If (_continueUpdateMultiple) Then
                            affectedRows = UpdateByBUAndDefaultName(businessUnit, System.Enum.GetName(GetType(TblEcommerceModuleDefaultsBuEnum), keyValue.Key), keyValue.Value, givenTransaction)
                            totalAffectedRows = totalAffectedRows + affectedRows
                        Else
                            Exit For
                        End If
                    Next
                End If
            End If
            'Return the results 
            Return totalAffectedRows
        End Function

#End Region

    End Class

End Namespace
